src/theory/bv/rewrites

```diff
@@ -0,0 +1,321 @@
+
+(define-rule bv-neg-mult
+  ((xs ?BitVec) (ys ?BitVec) (n Int) (m Int))
+  (bvneg (bvmul xs (@bv n m) ys))
+  (bvmul xs (@bv (- n) m) ys))
+(define-rule* bv-neg-add
+  ((x ?BitVec) (y ?BitVec) (zs ?BitVec :list))
+  (bvneg (bvadd x y zs))
+  (bvneg (bvadd y zs))
+  (bvadd (bvneg x) _))
+
+(define-rule bv-mult-distrib-const-neg
+  ((x ?BitVec) (n Int) (m Int))
+  (bvmul (bvneg x) (@bv n m))
+  (bvmul x (@bv (- n) m)))
+(define-rule bv-mult-distrib-const-add
+  ((x ?BitVec) (y ?BitVec) (n Int) (m Int))
+  (bvmul (bvadd x y) (@bv n m))
+  (bvadd
+    (bvmul x (@bv n m))
+    (bvmul y (@bv n m))
+  ))
+(define-rule bv-mult-distrib-const-sub
+  ((x ?BitVec) (y ?BitVec) (n Int) (m Int))
+  (bvmul (bvsub x y) (@bv n m))
+  (bvsub
+    (bvmul x (@bv n m))
+    (bvmul y (@bv n m))
+  ))
+
+(define-rule bv-mult-distrib-1
+  ((x1 ?BitVec) (x2 ?BitVec) (y ?BitVec))
+  (bvmul (bvadd x1 x2) y)
+  (bvadd (bvmul x1 y) (bvmul x2 y)))
+(define-rule bv-mult-distrib-2
+  ((x1 ?BitVec) (x2 ?BitVec) (y ?BitVec))
+  (bvmul y (bvadd x1 x2))
+  (bvadd (bvmul y x1) (bvmul y x2)))
+(define-rule bv-not-xor
+  ((x ?BitVec) (xs ?BitVec :list))
+  (bvnot (bvxor x xs))
+  (bvxor (bvnot x) xs))
+
+(define-rule* bv-and-simplify-1
+  ((xs ?BitVec :list) (ys ?BitVec :list) (zs ?BitVec :list) (x ?BitVec))
+  (bvand xs x ys x zs)
+  (bvand xs x ys zs))
+(define-rule* bv-and-simplify-2
+  ((xs ?BitVec :list) (ys ?BitVec :list) (zs ?BitVec :list) (x ?BitVec))
+  (bvand xs x ys (bvnot x) zs)
+  (@bv 0 (@bvsize x)))
+
+(define-rule* bv-or-simplify-1
+  ((xs ?BitVec :list) (ys ?BitVec :list) (zs ?BitVec :list) (x ?BitVec))
+  (bvor xs x ys x zs)
+  (bvor xs x ys zs))
+(define-rule* bv-or-simplify-2
+  ((xs ?BitVec :list) (ys ?BitVec :list) (zs ?BitVec :list) (x ?BitVec))
+  (bvor xs x ys (bvnot x) zs)
+  (bvnot (@bv 0 (@bvsize x))))
+
+(define-rule* bv-xor-simplify-1
+  ((xs ?BitVec :list) (ys ?BitVec :list) (zs ?BitVec :list) (x ?BitVec))
+  (bvxor xs x ys x zs)
+  (bvxor xs ys zs))
+(define-rule* bv-xor-simplify-2
+  ((xs ?BitVec :list) (ys ?BitVec :list) (zs ?BitVec :list) (x ?BitVec))
+  (bvxor xs x ys (bvnot x) zs)
+  (bvnot (bvxor xs ys zs)))
+(define-rule* bv-xor-simplify-3
+  ((xs ?BitVec :list) (ys ?BitVec :list) (zs ?BitVec :list) (x ?BitVec))
+  (bvxor xs (bvnot x) ys x zs)
+  (bvnot (bvxor xs ys zs)))
+
+; -- Simplify Bitblasting --
+
+; Unused
+; -a1 - a2 - ... - an + ak + ..  ==> - (a1 + a2 + ... + an) + ak
+;(define-rule bv-bb-add-neg )
+
+; x < y + 1 <=> (not y < x) and y != 1...1
+(define-cond-rule bv-ult-add-one
+  ((x ?BitVec) (y ?BitVec) (c1 ?BitVec))
+  (= c1 (@bv 1 (@bvsize c1)))
+  (bvult x (bvadd y c1))
+  (and
+    (not (bvult y x))
+    (not (= y (bvnot (@bv 0 (@bvsize y)))))))
+(define-cond-rule bv-concat-to-mult
+  ((x ?BitVec) (i Int) (m Int) (n0 Int))
+  (and (= (+ 1 i m) (@bvsize x)) (= n0 0))
+  (concat (extract i n0 x) (@bv n0 m))
+  (bvmul x (bvshl (@bv 1 (@bvsize x)) (@bv m (@bvsize x)))))
+
+; Currently bugged due to not being able to match n.
```

Comment:
Does bugged mean unused in this context? Or that there is a bug in this rule?

Reply:
The idea here was that if you have `(sign_extend n x)` and it matches (say #00001001), the rewriter would not be able to match `n`, and it could only match `n` if the input expression is written exactly as `(sign_extend n ...)`.

Reply:
RARE rules are independent from the C++ code even though we modeled bit-vector rules after that code. You could leave a comment saying that does not fully capture the rewrite done in the code but calling it buggy seems misleading if this is the only concern.  

I still don't quite understand the issue. Could you please elaborate a bit more.

"and it could only match n if the input expression is written exactly as (sign_extend n ...)"

it can not match `n` against a constant? Could you give me the whole term please? So it could match `(sign_extend n #b1001)` but not `(sign_extend 4 #b1001)`?

Reply:
I'll delete the comment. I think that comment was very misleading

Reply:
deleted